Trusted Types fails to protect against a script edited mid-parse

lukewarlow opened the pull request:
#### c31d9c40707d3ec58faee43f7446ac3fbd6c4531
<pre>
Trusted Types fails to protect against a script edited mid-parse
<a href="https://bugs.webkit.org/show_bug.cgi?id=274253">https://bugs.webkit.org/show_bug.cgi?id=274253</a>

Reviewed by Darin Adler.

This patch adds a new flag to script elements that is triggered if their children are changed by API.
This flag is then used to decide whether to accept parsed script elements as trusted.

This patch stops parsed scripts working in XHTML documents when TT is enforced. A follow up patch will address that.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLScriptElement-in-xhtml-document.tentative.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLScriptElement-internal-slot-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLScriptElement-internal-slot.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/SVGScriptElement-internal-slot-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/SVGScriptElement-internal-slot.html: Added.
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::childrenChanged):
(WebCore::ScriptElement::finishParsingChildren):
* Source/WebCore/dom/ScriptElement.h:

Canonical link: <a href="https://commits.webkit.org/279225@main">https://commits.webkit.org/279225@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/783b9e7d36e149de140dda4fb150458e6c7e2d48

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52782 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32119 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5268 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56060 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3505 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55087 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38801 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3228 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42846 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2262 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54880 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29837 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45561 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23959 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26928 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2876 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1664 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48803 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3029 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57652 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27922 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3049 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50243 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29143 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45755 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49521 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11544 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30059 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28896 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLScriptElement-internal-slot.html:
```diff
@@ -7,24 +7,17 @@
 <body>
 <script>
+  // This policy allows document.write to work as that's not relevant to what's actually being tested here.
+  trustedTypes.createPolicy('default', {createHTML: s => s});
```
Exported at https://github.com/web-platform-tests/wpt/pull/46308

Change is needed as the existing method was flaky and didn't reproduce properly in WebKit and Gecko.

darinadler commented on Source/WebCore/dom/ScriptElement.h:
```diff
@@ -157,6 +157,7 @@ class ScriptElement {
 
     // https://w3c.github.io/trusted-types/dist/spec/#slots-with-trusted-values
     String m_trustedScriptText { emptyString() };
+    bool m_childrenChangedByApi : 1 { false };
```
API should be kept all capitals, but not sure that "API" is what we want to say here, maybe "Script" is better. I guess we do use API for `ContainerNode::ChildChange::Source::API`, so we can stick with that if you like.

Writing `: 1` here is not a good optimization. This bit won’t be packed with other bits unless they are consecutive. So we should either move this up above with the other single bit field booleans, or we should omit the `: 1`.

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLScriptElement-in-xhtml-document.tentative.https-expected.txt:
```diff
@@ -1,3 +1,5 @@
+CONSOLE MESSAGE: This requires a TrustedScript value else it violates the following Content Security Policy directive: "require-trusted-types-for 'script'"
+
+Harness Error (TIMEOUT), message = null
```
We don't want tests in a state where they have to time out. Can we avoid this?

lukewarlow replied:
It's skipped in test expectations so it won't run. I'm working to make it so this test passes anyway but it requires a sizable changeset so doing it as followups.

lukewarlow committed:
Trusted Types fails to protect against a script edited mid-parse
https://bugs.webkit.org/show_bug.cgi?id=274253

Reviewed by Darin Adler.

This patch adds a new flag to script elements that is triggered if their children are changed by API.
This flag is then used to decide whether to accept parsed script elements as trusted.

This patch stops parsed scripts working in XHTML documents when TT is enforced. A follow up patch will address that.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLScriptElement-in-xhtml-document.tentative.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLScriptElement-internal-slot-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLScriptElement-internal-slot.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/SVGScriptElement-internal-slot-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/SVGScriptElement-internal-slot.html: Added.
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::childrenChanged):
(WebCore::ScriptElement::finishParsingChildren):
* Source/WebCore/dom/ScriptElement.h:

Canonical link: https://commits.webkit.org/279225@main
